// sites/node_modules/marshydro.js
let siteTargets = {
	"led growlights": {
		userAgent: "Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/72.0.3626.121 Safari/537.36",
		urls: [["https://www.mars-hydro.com/","led-grow-light","?","limit=25"]],
		createUrl(args){
			var ret = ''
			if(args.self){
				var args = args || {}
				var n = args.n || 0
				args.self.urls[n].forEach(part=>{
					ret += part
				})
			}
			return ret
		},
		list: {
			id: '#products-list'
		},
		"schema": {
			"name" : "thing",
			"props": {
				title: {
					selector: ".product-shop > .f-fix > .product-primary > .product-name > a",
					attribute: 'title',
					schema: {
						type: String,
						required: true
					}
				},
				subtitle: {
					selector: "",
					attribute: '',
					schema: {
						type: String,
					}
				},
				"short description": {
					selector: ".product-shop > .f-fix > .desc",
					attribute: '',
					html: true,
					schema: {
						type: String,
					}
				},
				price: {
					selector: ".product-shop > .f-fix > .old-price > .price",
					attribute: '',
					schema: {
						type: String,
						required: true
					}
				},
				"sale price": {
					selector: ".product-shop > .f-fix > .special-price > .price",
					attribute: '',
					html: true,
					schema: {
						type: String
					}
				},
				thumbnail: {
					selector: ".product-image > img",
					attribute: 'src',
					schema: {
						type: String,
						required: true
					}
				},
				images: {
					schema: {
						type: [String]
					}
				}
			}
		}
	}
}

module.exports = siteTargets
